rpc/app_querier.go

```diff
@@ -170,6 +308,53 @@ func (aq *AppQuerier) QueryLastValsetBeforeNonce(ctx context.Context, nonce uint
 	return resp.Valset, nil
 }
 
+// QueryHistoricalLastValsetBeforeNonce returns the last historical valset before nonce for a certain height.
+func (aq *AppQuerier) QueryHistoricalLastValsetBeforeNonce(ctx context.Context, nonce uint64, height uint64) (*celestiatypes.Valset, error) {
+	queryClient := celestiatypes.NewQueryClient(aq.clientConn)
+	var header metadata.MD
+	resp, err := queryClient.LatestValsetRequestBeforeNonce(
+		metadata.AppendToOutgoingContext(ctx, cosmosgrpc.GRPCBlockHeightHeader, strconv.FormatUint(height, 10)),
+		&celestiatypes.QueryLatestValsetRequestBeforeNonceRequest{Nonce: nonce},
+		grpc.Header(&header),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.Valset, nil
+}
+
+// QueryRecursiveHistoricalLastValsetBeforeNonce recursively looks for the last historical valset before nonce for a certain height until genesis.
+func (aq *AppQuerier) QueryRecursiveHistoricalLastValsetBeforeNonce(ctx context.Context, nonce uint64, height uint64) (*celestiatypes.Valset, error) {
```

Comment:
are both of these used independently, or do we only ever call one of these?

Reply:
One of them queries at an exact height, the other goes range by range of blocks to find the attestation. So yes, both is used